### PR TITLE
Lower TableGroupWithinPartitions

### PR DESCRIPTION
**Pull request by johnc1231**
This PR lowers `TableGroupWithinPartitions` and adds a test for it.

### CODE REVIEW

**Comment by johnc1231** [comment]
This isn't tested as thoroughly as I would like because I don't have a way to collect a keyed thing or unkey it in the lowerer.

**Comment by johnc1231** [comment]
Maybe ought to wait for #8649 to go in so I can add tests that depend on keying.

File: hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala [hunk 1]
```diff
@@ -224,6 +224,24 @@ object LowerTableIR {
             Subst(newRow, BindingEnv(env, scan = Some(env)))
           })
 
+        case TableGroupWithinPartitions(child, groupedStructName, n) =>
+          val loweredChild = lower(child)
+          val keyFields = FastIndexedSeq(loweredChild.partitioner.kType.fieldNames: _*)
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala:229]: I think the partitioner could have a longer key than the IR key. Use `child.typ.keyType`

Comment (johnc1231) [reply hail/src/main/scala/is/hail/expr/ir/lowering/LowerTableIR.scala]: Agreed and addressed

**Pull request merged**